File: .github/scripts/preflight.mjs
#!/usr/bin/env node
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { spawnSync } from 'node:child_process';

const root = path.resolve(import.meta.dirname, '../..');
const stacksDir = path.join(root, 'compose-stacks');
const namePattern = /^[a-z0-9][a-z0-9_-]*$/;
const mergeScript = path.join(root, '.github/scripts/merge-dotenv.mjs');

function fail(message) {
  console.error(`preflight: ${message}`);
  process.exit(1);
}
function command(program, args, options = {}) {
  const result = spawnSync(program, args, { encoding: 'utf8', ...options });
  if (result.error || result.status !== 0) {
    throw new Error(result.error?.message || result.stderr.trim() || `${program} exited with ${result.status}`);
  }
  return result.stdout;
}
function parseInventory() {
  let inventory;
  try {
    inventory = JSON.parse(command('ansible-inventory', ['--list', '-i', path.join(root, 'ansible/inventory.yaml')]));
  } catch (error) {
    fail(`could not read parsed Ansible inventory: ${error.message}`);
  }
  const hostvars = inventory._meta?.hostvars || {};
  const hosts = [];
  for (const hostname of inventory.workloads?.hosts || []) {
    const data = hostvars[hostname] || {};
    if (!(data.host_roles || []).includes('deploy')) continue;
    const id = hostname.split('.')[0];
    if (!namePattern.test(id) || id === 'docker-host') fail(`invalid or reserved deploy host identifier: ${id}`);
    hosts.push({ id, address: data.ansible_host || hostname });
  }
  if (!hosts.length) fail('no workload hosts have the deploy role');
  if (new Set(hosts.map((host) => host.id)).size !== hosts.length) fail('deploy host identifiers are not unique');
  return hosts.sort((a, b) => a.id.localeCompare(b.id));
}
function mergedEnv(shared, assignment) {
  try {
    return command('node', [mergeScript, shared, assignment]);
  } catch (error) {
    fail(`dotenv validation failed for ${assignment}: ${error.message}`);
  }
}
function validateCompose(stack, host, compose, envText) {
  const representative = envText.replace(/op:\/\/[^\s]+/g, 'example-secret');
  const envFile = path.join(os.tmpdir(), `compose-${stack}-${host}-${process.pid}.env`);
  fs.writeFileSync(envFile, representative, { mode: 0o600 });
  try {
    command('docker', ['compose', '--project-name', stack, '--env-file', envFile, '-f', compose, 'config', '--quiet'], { stdio: ['ignore', 'ignore', 'pipe'] });
  } catch (error) {
    fail(`Compose validation failed for ${stack}/${host}: ${error.message}`);
  } finally {
    fs.rmSync(envFile, { force: true });
  }
}
function main() {
  const hosts = parseInventory();
  const hostIds = new Set(hosts.map((host) => host.id));
  const stacks = [];
  for (const entry of fs.readdirSync(stacksDir, { withFileTypes: true })) {
    if (!entry.isDirectory() || entry.name === 'docker-host') continue;
    const stack = entry.name;
    const stackDir = path.join(stacksDir, stack);
    if (!namePattern.test(stack)) fail(`invalid stack name: ${stack}`);
    const compose = path.join(stackDir, 'docker-compose.yaml');
    if (!fs.existsSync(compose)) fail(`application stack ${stack} has no docker-compose.yaml`);
    const deployments = path.join(stackDir, 'deployments');
    const shared = path.join(deployments, '_shared.env');
    if (fs.existsSync(shared) && !fs.statSync(shared).isFile()) fail(`${shared} is not a regular file`);
    if (fs.existsSync(deployments)) {
      for (const assignmentEntry of fs.readdirSync(deployments, { withFileTypes: true })) {
        if (assignmentEntry.name === '_shared.env') continue;
        const assignment = path.join(deployments, assignmentEntry.name);
        if (!assignmentEntry.isFile() || !assignmentEntry.name.endsWith('.env')) fail(`invalid deployment file: ${assignment}`);
        const host = assignmentEntry.name.slice(0, -4);
        if (!namePattern.test(host) || host === 'docker-host') fail(`invalid deployment host filename: ${assignmentEntry.name}`);
        if (!hostIds.has(host)) fail(`unknown deployment host ${host} in ${assignment}`);
        validateCompose(stack, host, compose, mergedEnv(shared, assignment));
      }
    }
    stacks.push(stack);
  }
  if (!stacks.length) fail('no application Compose stacks found');
  const output = { hosts, stacks: stacks.sort() };
  if (process.env.GITHUB_OUTPUT) {
    fs.appendFileSync(process.env.GITHUB_OUTPUT, `hosts=${JSON.stringify(output.hosts)}\nstacks=${JSON.stringify(output.stacks)}\n`);
  }
  console.log(JSON.stringify(output));
}
main();

File: .github/scripts/preflight.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'json'
require 'open3'
require 'tempfile'

ROOT = File.expand_path('../..', __dir__)
STACKS_DIR = File.join(ROOT, 'compose-stacks')
NAME_PATTERN = /\A[a-z0-9][a-z0-9_-]*\z/
DISCOVER_SCRIPT = File.join(ROOT, '.github/scripts/discover-deploy-hosts.rb')
MERGE_SCRIPT = File.join(ROOT, '.github/scripts/merge-dotenv.rb')

def fail_preflight(message)
  warn "preflight: #{message}"
  exit 1
end

def command(program, *args, **options)
  stdout, stderr, status = Open3.capture3(program, *args, **options)
  raise(stderr.strip.empty? ? "#{program} exited with #{status.exitstatus}" : stderr.strip) unless status.success?
  stdout
end

def parse_inventory
  JSON.parse(command('ruby', DISCOVER_SCRIPT, File.join(ROOT, 'ansible/inventory.yaml')))
rescue StandardError => e
  fail_preflight("could not read deploy hosts: #{e.message}")
end

def merged_env(shared, assignment)
  command('ruby', MERGE_SCRIPT, shared, assignment)
rescue StandardError => e
  fail_preflight("dotenv validation failed for #{assignment}: #{e.message}")
end

def validate_compose(stack, host, compose, env_text)
  representative = env_text.gsub(/op:\/\/[^\s]+/, 'example-secret')
  env_file = Tempfile.new(["compose-#{stack}-#{host}-", '.env'])
  env_file.write(representative)
  env_file.flush
  env_file.chmod(0o600)
  begin
    command('docker', 'compose', '--project-name', stack, '--env-file', env_file.path, '-f', compose, 'config', '--quiet', err: [:child, :out])
  rescue StandardError => e
    fail_preflight("Compose validation failed for #{stack}/#{host}: #{e.message}")
  ensure
    env_file.close!
  end
end

hosts = parse_inventory
host_ids = hosts.to_h { |host| [host['id'], true] }
stacks = []

Dir.each_child(STACKS_DIR) do |stack|
  stack_dir = File.join(STACKS_DIR, stack)
  next unless File.directory?(stack_dir) && stack != 'docker-host'

  fail_preflight("invalid stack name: #{stack}") unless NAME_PATTERN.match?(stack)
  compose = File.join(stack_dir, 'docker-compose.yaml')
  fail_preflight("application stack #{stack} has no docker-compose.yaml") unless File.file?(compose)

  deployments = File.join(stack_dir, 'deployments')
  shared = File.join(deployments, '_shared.env')
  fail_preflight("#{shared} is not a regular file") if File.exist?(shared) && !File.file?(shared)

  if File.directory?(deployments)
    Dir.each_child(deployments) do |filename|
      next if filename == '_shared.env'

      assignment = File.join(deployments, filename)
      fail_preflight("invalid deployment file: #{assignment}") unless File.file?(assignment) && filename.end_with?('.env')
      host = filename.delete_suffix('.env')
      fail_preflight("invalid deployment host filename: #{filename}") unless NAME_PATTERN.match?(host) && host != 'docker-host'
      fail_preflight("unknown deployment host #{host} in #{assignment}") unless host_ids.key?(host)
      validate_compose(stack, host, compose, merged_env(shared, assignment))
    end
  end
  stacks << stack
end

fail_preflight('no application Compose stacks found') if stacks.empty?
output = { 'hosts' => hosts, 'stacks' => stacks.sort }
if (github_output = ENV['GITHUB_OUTPUT'])
  File.open(github_output, 'a') do |file|
    file.puts "hosts=#{JSON.generate(output['hosts'])}"
    file.puts "stacks=#{JSON.generate(output['stacks'])}"
  end
end
puts JSON.generate(output)
